### PR TITLE
feat: add throwICUArgumentMismatch option to throw on ICU argument mismatch

### PR DESCRIPTION
**Pull request by makepanic**
This implements ICU argument checking to solve #601.

I have a couple open questions:

* is `throwICUArgumentMismatch` ok as a name?
* should this be added to the docs?
* is it ok if this PR introduces a [mapping an array with side-effects](https://github.com/ember-intl/ember-intl/compare/master...makepanic:issues/601?expand=1#diff-8218ca25365bb38750c0455c805b9fbeR110)?
* should this also support a configurable filter function to allow ICU mismatch in some cases (similar to `requiresTranslation`)?

### CODE REVIEW

File: tests-node/unit/broccoli/translation-reducer-test.js [hunk 1]
```diff
@@ -7,6 +7,14 @@ const subject = require('../../../lib/broccoli/translation-reducer');
 
 describe('translation-reducer', function() {
   beforeEach(function() {
+    this.icuFixture = {
+      de: {
+        foo: 'Hallo {who}',
```
Comment (jasonmit) [inline tests-node/unit/broccoli/translation-reducer-test.js]: Can you write a test that uses parameters.

`You have {numPhotos, plural, =0 {no photos.} =1 {one photo.} other {# photos.}}`

I'd expect it to not fail if the parameters are different, only should match on first param (key).

File: index.js [hunk 2]
```diff
@@ -25,6 +25,7 @@ const defaultConfig = {
   autoPolyfill: false,
   inputPath: 'translations',
   outputPath: 'translations',
+  throwICUArgumentMismatch: false,
```
Comment (jasonmit) [inline index.js]: `throwICUArgumentMismatch` -> `errorOnNamedArgumentMismatch`

(Context: I want to go back and rename `throwMissingTranslations` to `errorOnMissingTranslations`)

**Comment by jasonmit** [comment]
> is it ok if this PR introduces a mapping an array with side-effects?

I'm fine with that.

> should this be added to the docs?

I think the ember-intl config file itself can act as its own doc via comments for now. 

> should this also support a configurable filter function to allow ICU mismatch in some cases (similar to requiresTranslation)?

I wouldn't make that a requirement for this to land.

**Comment by makepanic** [comment]
I've updated the PR

File: tests-node/unit/broccoli/translation-reducer-test.js [hunk 3]
```diff
@@ -7,6 +7,18 @@ const subject = require('../../../lib/broccoli/translation-reducer');
...
+        photos: 'You have {numPhotos, plural, =0 {no photos.} =1 {one photo.} other {# photos.}}',
```
Comment (jasonmit) [inline tests-node/unit/broccoli/translation-reducer-test.js]: Can you add a test that this is valid:

```
en: {
  photos: 'You have {numPhotos, plural, =0 {no photos.} =1 {one photo.} other {# photos.}}'
},
de: {
  photos: 'You have {numPhotos, plural, =0 {foo} =1 {bar} other {# baz}}'
}
```

File: lib/broccoli/translation-reducer.js [hunk 4]
```diff
@@ -82,30 +82,80 @@ class TranslationReducer extends CachingWriter {
...
 
-  findMissingTranslations(translations) {
+  extractICUArguments(string) {
+    return parser
+      .parse(string)
+      .elements.filter(element => element.type === 'argumentElement')
+      .map(argumentElement => argumentElement.id);
+  }
+
+  lintTranslations(translations) {
```
Comment (jasonmit) [inline lib/broccoli/translation-reducer.js]: Can we break this function up into smaller fns?

**Comment by makepanic** [comment]
I've added the valid argument test and tried to break the "linting" into smaller methods.
I've also rebased onto current master.

**Comment by jasonmit** [comment]
Thanks again @makepanic :)  It will be in the next 4.x beta

**Pull request merged**